### PR TITLE
Fix SSE hang on close

### LINKED ISSUE #526: Fix and unskip flaky test: SSETransportStopsSendAndReceiveLoopsWhenTransportStopped
We're skipping it for now to pass the CI, but we should give it an actual investigation.

### PR DESCRIPTION
**Pull request by BrennanConroy**
#526 
@moozzyk @davidfowl

### CODE REVIEW

File: src/Microsoft.AspNetCore.Sockets.Client.Http/ServerSentEventsTransport.cs [hunk 1]
```diff
@@ -80,7 +81,14 @@ public Task StartAsync(Uri url, Channel<byte[], SendMessage> application, Transf
             var response = await _httpClient.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, cancellationToken);
 
             var stream = await response.Content.ReadAsStreamAsync();
-            var pipelineReader = stream.AsPipelineReader(cancellationToken);
+            var pipe = _pipelineFactory.Create();
+
+            _ = Task.Run(async () =>
```
Comment (moozzyk) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/ServerSentEventsTransport.cs]: Why is this needed?

File: src/Microsoft.AspNetCore.Sockets.Client.Http/ServerSentEventsTransport.cs [hunk 2]
```diff
@@ -132,6 +137,16 @@ public Task StartAsync(Uri url, Channel<byte[], SendMessage> application, Transf
                 _transportCts.Cancel();
                 stream.Dispose();
                 _logger.ReceiveStopped(_connectionId);
+                pipe.Writer.Complete();
+
+                try
+                {
+                    await writeTask;
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/ServerSentEventsTransport.cs]: Should we handle cases where this task takes too long/never finishes, or is that not something that happens except in test code

File: src/Microsoft.AspNetCore.Sockets.Client.Http/ServerSentEventsTransport.cs [hunk 3]
```diff
@@ -140,6 +137,16 @@ public Task StartAsync(Uri url, Channel<byte[], SendMessage> application, Transf
...
+                }
+                catch (Exception ex)
```
Comment (moozzyk) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/ServerSentEventsTransport.cs]: I looked at code [here](https://github.com/dotnet/corefxlab/blob/master/src/System.IO.Pipelines.Extensions/StreamExtensions.cs#L71-L83) and it seems that exception handling is actually not needed. Also, the writer will be completed so it seems that the only thing to do is to await the task...

File: src/Microsoft.AspNetCore.Sockets.Client.Http/ServerSentEventsTransport.cs [hunk 4]
```diff
@@ -21,6 +21,7 @@ public class ServerSentEventsTransport : ITransport
         private readonly ILogger _logger;
         private readonly CancellationTokenSource _transportCts = new CancellationTokenSource();
         private readonly ServerSentEventsMessageParser _parser = new ServerSentEventsMessageParser();
+        private readonly PipeFactory _pipelineFactory = new PipeFactory();
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/ServerSentEventsTransport.cs]: Who disposes this?

File: src/Microsoft.AspNetCore.Sockets.Client.Http/ServerSentEventsTransport.cs [hunk 5]
```diff
@@ -132,6 +137,16 @@ public Task StartAsync(Uri url, Channel<byte[], SendMessage> application, Transf
                 _transportCts.Cancel();
                 stream.Dispose();
                 _logger.ReceiveStopped(_connectionId);
+                pipe.Writer.Complete();
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/ServerSentEventsTransport.cs]: Seems likt CopyToEndAsync already completes the writer

Comment (davidfowl) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/ServerSentEventsTransport.cs]: https://github.com/dotnet/corefxlab/blob/master/src/System.IO.Pipelines.Extensions/StreamExtensions.cs#L71-L83

Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/ServerSentEventsTransport.cs]: Is it safe to assume that the stream will end and complete will actually be called? In our tests the wait for `writeTask` hangs because the stream is constantly being written to.

Comment (davidfowl) [inline src/Microsoft.AspNetCore.Sockets.Client.Http/ServerSentEventsTransport.cs]: I see, so if we abort, this will complete the pipe writer

**Comment by muratg** [comment]
@BrennanConroy *ping*

**Comment by BrennanConroy** [comment]
Blocked on 2.1 runtime
________________________________

**Comment by muratg** [comment]
OK thanks. @natemcmaster will be looking into bringing 2.1 runtime soon.

**Comment by davidfowl** [comment]
@BrennanConroy whats the deal with this?

**Comment by BrennanConroy** [comment]
Made a different PR https://github.com/aspnet/SignalR/pull/1279
Not happy with either one though

**Comment by davidfowl** [comment]
Ok closing this
